ICRCs/ICRC-37/ICRC-37.md

```diff
@@ -0,0 +1,476 @@
+    memo : opt blob;
+    created_at_time : opt nat64;
+};
+
+type RevokeCollectionApprovalResult = variant {
+    Ok : nat; // Transaction index for successful revocation
+    Err : RevokeCollectionApprovalError;
+};
+
+type RevokeCollectionApprovalError = variant {
+    ApprovalDoesNotExist;
+    TooOld;
+    CreatedInFuture : record { ledger_time: nat64 };
+    GenericError : record { error_code : nat; message : text };
+    BatchTermination;
+    GenericBatchError : record { error_code : nat; message : text };
+};
+```
+
+```candid "Methods" +=
+icrc37_revoke_collection_approvals: (vec RevokeCollectionApprovalArg)
+    -> (vec opt RevokeCollectionApprovalResult);
+```
+
+### icrc37_is_approved
+
+Returns `true` if an active approval, i.e., a token-level approval or collection-level approval, exists that allows the `spender` to transfer the token `token_id` from the given `from_subaccount`, `false` otherwise.
+
+```candid "Type definitions" +=
+type IsApprovedArg = record {
+    spender : Account;
+    from_subaccount : opt blob;
+    token_id : nat;
+};
+```
+
+```candid "Methods" +=
+icrc37_is_approved : (vec IsApprovedArg)
+    -> (vec bool) query;
+```
+
+### icrc37_get_token_approvals
+
+Returns the token-level approvals that exist for the given vector of `token_ids`.  The result is paginated, the mechanics of pagination are analogous to `icrc7_tokens` using `prev` and `take` to control pagination, with `prev` being of type `TokenApproval`. Note that `take` refers to the number of returned elements to be requested. The `prev` parameter is a `TokenApproval` element with the meaning that `TokenApproval`s following the provided one are returned, based on a sorting order over `TokenApproval`s implemented by the ledger.
+
+The response is a vector of `TokenApproval` elements. If multiple approvals exist for a token id, multiple entries of type `TokenApproval` with the same token id are contained in the response.
+
+The ordering of the elements in the response is undefined. An implementation of the ledger can use any internal sorting order for the elements of the response to implement pagination.
+
+See the [#icrc37_approve_tokens](#icrc37_approve_tokens) for the `ApprovalInfo` type.
+
+```candid "Type definitions" +=
+type TokenApproval = record {
+    token_id : nat;
+    approval_info : ApprovalInfo;
+};
+```
+
+```candid "Methods" +=
+icrc37_get_token_approvals : (token_ids : vec nat, prev : opt TokenApproval; take : opt nat)
+    -> (vec TokenApproval) query;
+```
+
+> [!NOTE] This method deviates from the API best practice outlined in ICRC-7 of not having paginated batch APIs. The reason is that this method requires pagination because of possibly large numbers of responses, but is also expected to be useful as batch method for frequently expected use cases. Thus, the methods cannot use the typical positional arguments of other batch calls in the ICRC-7 and ICRC-37 standards.
```

Comment:
```suggestion
> [!NOTE] 
> This method deviates from the API best practice outlined in ICRC-7 of not having paginated batch APIs. The reason is that this method requires pagination because of possibly large numbers of responses, but is also expected to be useful as batch method for frequently expected use cases. Thus, the methods cannot use the typical positional arguments of other batch calls in the ICRC-7 and ICRC-37 standards.
```

Reply:
Applied.